File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        int answer = 0;
        long sum1 = sumQ(queue1);
        long sum2 = sumQ(queue2);
        long total = sum1 + sum2;

        if (total % 2 != 0) return -1; // 전체 합이 홀수면 불가능

        long target = total / 2;

        Queue<Integer> q1 = new LinkedList<>();
        Queue<Integer> q2 = new LinkedList<>();
        for (int i = 0; i < queue1.length; i++) {
            q1.add(queue1[i]);
            q2.add(queue2[i]);
        }

        int maxOperations = queue1.length * 3; // 안전한 최대 이동 횟수

        while (answer <= maxOperations) {
            if (sum1 == target) return answer;
            else if (sum1 > target) {
                int v = q1.poll();
                sum1 -= v;
                sum2 += v;
                q2.add(v);
            } else {
                int v = q2.poll();
                sum2 -= v;
                sum1 += v;
                q1.add(v);
            }
            answer++;
        }

        return -1; // 무한 루프 방지
    }

    private static int sumQ(int[] q) {
        int sum = 0;
        for (int val : q) {
            sum += val;
        }
        return sum;
    }
}

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.js
function solution(queue1, queue2) {
    let answer = 0;

    let sum1 = queue1.reduce((a, b) => a + b, 0);
    let sum2 = queue2.reduce((a, b) => a + b, 0);
    const total = sum1 + sum2;

    if (total % 2 !== 0) return -1; // 전체 합이 홀수면 불가능

    const target = total / 2;
    const max = queue1.length * 3;

    // queue1, queue2를 진짜 큐처럼 쓰기 위해 index로 관리
    let q = [...queue1, ...queue2];
    let i = 0;  // queue1의 시작
    let j = queue1.length;  // queue2의 시작

    while (answer <= max) {
        if (sum1 === target) return answer;

        if (sum1 > target) {
            const value = q[i++];
            sum1 -= value;
            sum2 += value;
        } else {
            const value = q[j++];
            sum2 -= value;
            sum1 += value;
        }

        answer++;
    }

    return -1;
}
